apps/labs/posts/array-api-support-scikit-learn.md

```diff
@@ -0,0 +1,149 @@
+
+With the release of scikit-learn 1.3, we enabled experimental Array API support for a limited set of machine learning models. Array API support is gradually expanding to include more machine learning models and functionality on the development branch. Scikit-learn depends on the `array_api_compat` library for Array API support. `array_api_compat` extends the Array API standard to the main namespaces of NumPy's arrays, CuPy's arrays, and PyTorch's tensors. In this blog post, we cover scikit-learn's public interface for enabling Array API, the performance gain of running on an accelerator, and the challenges we faced when integrating Array API.
+
+## Benchmarks
+
+Scikit-learn was initially developed to run on CPUs with NumPy arrays. With Array API support, a limited set of scikit-learn models and tools can now run with other array libraries and devices like GPUs. The following benchmark results are from running scikit-learn's `LinearDiscriminantAnalysis` with different array libraries:
+
+![Bar plot with benchmark results comparing NumPy, PyTorch on CPU, and PyTorch on NVIDIA 3090 running Linear Discriminant Analysis. The PyTorch GPU results are marked as 27 times faster compared to NumPy for fitting the model and 28 times faster compared to NumPy for prediction.](/posts/array-api-support-scikit-learn/benchmark_results.png)
+
+The training and prediction times are improved when using PyTorch compared to NumPy. Running the computation on PyTorch CPU tensors is faster than NumPy because PyTorch CPU operations are multi-threaded by default.
+
+## scikit-learn's Array API interface
+
+Scikit-learn extended its experimental Array API support in version 1.3 to support NumPy's `ndarrays`, CuPy's `ndararys`, and PyTorch `Tensors`. By themselves, these array objects do not implement the Array API specification. To overcome this limitation, Quansight engineer Aaron Meurer, led the development of `array_api_compat` to adopt the Array API to NumPy, CuPy, and PyTorch. Scikit-learn directly uses `array_api_compat` for its Array API support. There are two APIs enabling Array API in scikit-learn: a global configuration and a context manager. The following example uses a context manager:
+
+```python
+import sklearn
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+from sklearn.datasets import make_classification
+import torch
+
+X_np, y_np = make_classification(random_state=0, n_samples=500_000, n_features=300)
+X_torch_cpu, y_torch_cpu = torch.asarray(X_np), torch.asarray(y_np)
+
+with sklearn.config_context(array_api_dispatch=True):
+    lda_torch_cpu = LinearDiscriminantAnalysis()
+    lda_torch_cpu.fit(X_torch_cpu, y_torch_cpu)
+
+    predictions = lda_torch_cpu.predict(X_torch_cpu)
```

Comment:
This example runs now, but it was slightly annoying that it required installing `array-api-compat` from its `main` branch (it needed `xp.take` to have the `axis` keyword be optional). I'll ask Aaron to cut a new release, so that things work out of the box with a `pip install array-api-compat`.

Reply:
Actually, let's do that here. @asmeurer would you mind cutting a new release of `array-api-compat`?

Reply:
@asmeurer could you upload a release by hand if the auto-publish thing is too messy? Should take only a few minutes for a pure Python package I'd think. It'd be great if this release was up by tomorrow.

Reply:
The release is out. I don't know if you need conda-forge too, but that should be out as soon as the bot catches it.

Reply:
Thanks Aaron! conda-forge would be nice, but less urgent. let's let the bot do its thing